bip-csfs.md

```diff
@@ -0,0 +1,124 @@
+```
+BIP: XXX
+Layer: Consensus (soft fork)
+Title: CHECKSIGFROMSTACK
+Author: Brandon Black <freedom@reardencode.com>, Jeremy Rubin <j@rubin.io>
+Comments-URI: https://github.com/bitcoin/bips/wiki/Comments:BIP-XXXX
+Status: Draft
+Type: Standards Track
+Created: 2023-12-22
+License: BSD-3-CLAUSE
+```
+
+## Abstract
+
+This BIP describes two new opcodes for the purpose of checking cryptographic
+signatures in bitcoin scripts against data other than bitcoin transactions.
+
+## Summary
+
+We propose replacing `OP_NOP5` (0xb4) in bitcoin script with
+`OP_CHECKSIGFROMSTACKVERIFY`. When verifying taproot script spends having
+leaf version 0xc0 (as defined in [BIP 342]), we propose `OP_CHECKSIGFROMSTACK`
+to replace `OP_SUCCESS204` (0xcc).
+
+`OP_CHECKSIGFROMSTACK` and `OP_CHECKSIGFROMSTACKVERIFY`
+have semantics similar to `OP_CHECKSIG` and
+`OP_CHECKSIGVERIFY` respectively, as specified below.
+
+Only 32-byte keys are constrained. Similar to [BIP 341] unknown key types, for
+other key lengths no signature verification is performed.
+
+## Specification
+
+* If fewer than 3 elements are on the stack, the script MUST fail and terminate immediately.
+* The public key (top element), message (second to top element), and signature (third from top element) are read from the stack.
```

Comment:
nit: I'd like to offer the suggestion that the push order goes:

`<msg> <sig> <key>` instead of `<sig> <msg> <key>`. The current `CHECKSIG` ordering is `<sig> <key>` and it seems odd to me that we would make the CSFS opcode put the msg *between* these other pushes. Alternatively you could go `<sig> <key> <msg>` and it would seem to me as equally intuitive. As is though it feels like a gotcha.

Is there a specific reason this choice was made that I hadn't considered?

Reply:
Anything that comes from the input stack should be before things that are provided by the locking script, so `sig` definitely comes first. Depending on particulars `msg` might come from the input stack or the locking script, and `key` always comes from the locking script. Examples:
* `<delegate-sig> <sig> <delegate-key> || DUP TOALT <key> CSFS FROMALT CHECKSIG`
* `<sig> <hash> || CTV <key> CSFS`
* `<sig> || <flags> TXHASH <key> CSFS`
* `<sig> <settlement-hash> <update-hash> || CTV SHA256 CAT <key> CSFS`

Reply:
Retracted.

I understand the general direction of what you're getting at but I have to admit that I'm not following the specific use cases where `<msg>` might come from the witness instead of the script. It makes sense to me why you'd want `<msg>` in the script, what's lest clear is under what conditions it is useful to let the witness define it.

Reply:
In any kind of delegation (see the first example above) the message is provided by the witness (i.e. the thing you're delegating to by signing). Here's a powerful example:

`<delegate_sig> <sig> <delegate_pubkey> <locktime> || CLTV SHA256 CAT <pubkey> CSFS VERIFY CHECKSIG`

Delegate to a key but only after `<locktime>`, with different locktimes per delegate key.

Reply:
Forgive me but I have to expand this out because it doesn't make sense to me:

Witness:
1. delegate signature is pushed onto the stack: `<delegate_sig>`
2. the signature authorizing the delegate is pushed: `<delegate_sig> <authorizing_sig>`
3. the delegate pubkey is pushed: `<delegate_sig> <authorizing_sig> <delegate_pubkey>`
4. locktime is pushed: `<delegate_sig> <authorizing_sig> <delegate_pubkey> <locktime>`

Script:
5. locktime is checked: `<delegate_sig> <authorizing_sig> <delegate_pubkey> <locktime>`
6. locktime is hashed: `<delegate_sig> <authorizing_sig> <delegate_pubkey> <locktime_hash>`
7. locktime hash combined with delegate pubkey: `<delegate_sig> <authorizing_sig> <delegate_pubkey || locktime_hash>`
8. authorizing pubkey is pushed: `<delegate_sig> <authorizing_sig> <delegate_pubkey || locktime_hash> <authorizing_pubkey>`
9. CSFS(V) operation(s) checks that the delegate pubkey is authorized after the locktime by checking the `key || locktime` message was signed by `authorizing_sig`: `<delegate_sig>`
10. CHECKSIG operation fails because we don't have a key? Maybe you forgot a dup somewhere? Maybe I don't understand script as well as I thought? idk.

Assuming this is roughly what you meant, then I'd imagine in this case the authorizing signature would be pre-signed and distributed. Then at transaction time the delegate presents the authorizing signature along with their key and assigned locktime.

Assuming all of this is a fair reading, then I think I get what you mean with `<msg>` having an interesting use case wherein it could be:
1. provided by the script
2. provided by the witness
3. collaboratively derived from witness and script participation.

... which seems pretty cool. I'd love if you can correct either my understanding or your example for completeness, but otherwise I'm satisfied with this and this thread can be resolved.

Reply:
Sorry for the long delay. You're quite correct that the example I gave does not work. Apparently I should have waited for Taproot Wizards new neovim script plugin. I believe this works for 3 more opcodes :(

`<sig> <delegate_sig> <delegate_pubkey> <locktime> || CLTV SHA256 OVER CAT TOALT CHECKSIGVERIFY FROMALT <pubkey> CSFS`

1. the signature authorizing the delegate is pushed: `<authorizing_sig>`
2. delegate signature is pushed onto the stack: `<authorizing_sig> <delegate_sig>`
3. the delegate pubkey is pushed: `<authorizing_sig> <delegate_sig> <delegate_pubkey>`
4. locktime is pushed: `<authorizing_sig> <delegate_sig> <delegate_pubkey> <locktime>`
5. locktime is checked: `<authorizing_sig> <delegate_sig> <delegate_pubkey> <locktime>`
6. locktime is hashed: `<authorizing_sig> <delegate_sig> <delegate_pubkey> <locktime_hash>`
7. delegate pubkey is copied over: `<authorizing_sig> <delegate_sig> <delegate_pubkey> <locktime_hash> <delegate_pubkey>`
8. locktime hash combined with delegate pubkey: `<authorizing_sig> <delegate_sig> <delegate_pubkey> <locktime_hash || delegate_pubkey>`
9. concatenation is moved to alt stack: `<authorizing_sig> <delegate_sig> <delegate_pubkey>`
10. CHECKSIGVERIFY validates the signature from the delegated key: `<authorizing_sig>`
11. concatenation is moved from alt stack: `<authorizing_sig> <locktime_hash || delegate_pubkey>`
12. authorizing pubkey is pushed: `<authorizing_sig> <locktime_hash || delegate_pubkey> <authorizing_pubkey>`
10. CSFS operation(s) checks that the delegate pubkey is authorized after the locktime by checking the `locktime || key` message was signed by `authorizing_sig`: `TRUE`

Thanks for checking the script.